Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Factory.swift

```diff
@@ -0,0 +1,356 @@
+        case .http_unix, .unix:
+            return bootstrap.connect(unixDomainSocketPath: self.key.unixPath)
+        case .https, .https_unix:
+            preconditionFailure("Unexpected scheme")
+        }
+    }
+
+    private func makeHTTPProxyChannel(
+        _ proxy: HTTPClient.Configuration.Proxy,
+        connectionID: HTTPConnectionPool.Connection.ID,
+        eventLoop: EventLoop,
+        logger: Logger
+    ) -> EventLoopFuture<(Channel, HTTPVersion)> {
+        // A proxy connection starts with a plain text connection to the proxy server. After
+        // the connection has been established with the proxy server, the connection might be
+        // upgraded to TLS before we send our first request.
+        let bootstrap = self.makePlainBootstrap(eventLoop: eventLoop)
+        return bootstrap.connect(host: proxy.host, port: proxy.port).flatMap { channel in
+            let encoder = HTTPRequestEncoder()
+            let decoder = ByteToMessageHandler(HTTPResponseDecoder(leftOverBytesStrategy: .dropBytes))
+            let proxyHandler = HTTP1ProxyConnectHandler(
+                targetHost: self.key.host,
+                targetPort: self.key.port,
+                proxyAuthorization: proxy.authorization,
+                deadline: .now() + .seconds(10)
+            )
+
+            do {
+                try channel.pipeline.syncOperations.addHandler(encoder)
+                try channel.pipeline.syncOperations.addHandler(decoder)
+                try channel.pipeline.syncOperations.addHandler(proxyHandler)
+            } catch {
+                return channel.eventLoop.makeFailedFuture(error)
+            }
+
+            return proxyHandler.proxyEstablishedFuture.flatMap {
+                channel.pipeline.removeHandler(proxyHandler).flatMap {
+                    channel.pipeline.removeHandler(decoder).flatMap {
+                        channel.pipeline.removeHandler(encoder)
+                    }
+                }
+            }.flatMap {
+                self.setupTLSInProxyConnectionIfNeeded(channel, logger: logger)
+            }
+        }
+    }
+
+    private func makeSOCKSProxyChannel(
+        _ proxy: HTTPClient.Configuration.Proxy,
+        connectionID: HTTPConnectionPool.Connection.ID,
+        eventLoop: EventLoop,
+        logger: Logger
+    ) -> EventLoopFuture<(Channel, HTTPVersion)> {
+        // A proxy connection starts with a plain text connection to the proxy server. After
+        // the connection has been established with the proxy server, the connection might be
+        // upgraded to TLS before we send our first request.
+        let bootstrap = self.makePlainBootstrap(eventLoop: eventLoop)
+        return bootstrap.connect(host: proxy.host, port: proxy.port).flatMap { channel in
+            let socksConnectHandler = SOCKSClientHandler(targetAddress: .domain(self.key.host, port: self.key.port))
+            let socksEventHandler = SOCKSEventsHandler(deadline: .now() + .seconds(10))
```

Comment:
I'd recommend adding a small comment re why 10 seconds?

Reply:
Refactored the magic numbers into their own struct:
```swift
        struct Timeouts {
            /// the `TimeAmount` waited before the TCP connection creation is failed with a timeout error
            var connect: TimeAmount = .seconds(10)

            /// the `TimeAmount` waited before the SOCKS proxy connection creation is failed with a timeout error
            var socksProxyHandshake: TimeAmount = .seconds(10)

            /// the `TimeAmount` waited before the HTTP proxy connection creation is failed with a timeout error
            var httpProxyHandshake: TimeAmount = .seconds(10)

            /// the `TimeAmount` waited before we the TLS handshake is failed with a timeout error
            var tlsHandshake: TimeAmount = .seconds(10)
        }
```
The 10 seconds are totally arbitrary. We need a timeout to not fill up our ConnectionPool with starting connections that never come up. Eventually we could make those publicly changeable.